File: companies-leetcode/facebook/linked-list/flatten-binary-tree-to-linked-list.js
/*

*/
// 1st attempt: 83rd percentile. Cleaner than faster implementations. Only 
// a few ms out of ~65 slower. Same time complexity.
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    if (!root) return null;
    let stack = [],
        prev = new TreeNode(-1)
    stack.push(root);
    
    while (stack.length>0) {
        let cur = stack.pop();
        if (cur && cur.right) stack.push(cur.right);
        if (cur && cur.left) stack.push(cur.left);
        
        prev.right = cur;
        prev.left = null;
        prev = prev.right;
    }
};

File: companies-leetcode/facebook/linked-list/flatten-binary-tree-to-linked-list.py
"""
Given a binary tree, flatten it to a linked list in-place.

For example, given the following tree:

    1
   / \
  2   5
 / \   \
3   4   6
The flattened tree should look like:

1
 \
  2
   \
    3
     \
      4
       \
        5
         \
          6
"""

# 1st attempt: 100th percentile. Iterative. Standard pre order dfs.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root: 
            return None
        stack = [root]
        prev = TreeNode(-1)
        
        while stack:
            cur = stack.pop()
            
            for c in [cur.right, cur.left]:
                if c:
                    stack.append(c)
            
            prev.right = cur
            prev.left = None
            prev = prev.right


"""
Notes:

Iteative pre order dfs...
Process the current node. Put the children on the stack in the opposite order that you want to process.
Say you put left on last. Then you'll always be popping off a left node until you've gon all the way down to the
right. Then will go back to the right nodes when we're out of left.

As for this specific problem...At each step we want the last node that was processed. So we just
keep track of a previous. 
"""